File: backend/app/schemas/financial.py
from decimal import Decimal
from enum import StrEnum

from pydantic import BaseModel, Field


class RiskLevel(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class MonthSummary(BaseModel):
    year: int
    month: int
    salary: Decimal
    total_expenses: Decimal
    monthly_installment_commitment: Decimal
    total_committed: Decimal
    remaining_balance: Decimal
    committed_percentage: Decimal
    active_installments: int


class FutureMonthBalance(BaseModel):
    months_ahead: int
    year: int
    month: int
    projected_installment_commitment: Decimal
    projected_balance: Decimal
    active_installments: int


class FutureBalanceResponse(BaseModel):
    salary: Decimal
    months: list[FutureMonthBalance]


class CanIBuyRequest(BaseModel):
    model_config = {
        "json_schema_extra": {
            "example": {
                "product_price": 3500,
                "installments": 10,
                "product_name": "New laptop",
            }
        }
    }

    product_price: Decimal = Field(gt=0, decimal_places=2)
    installments: int = Field(gt=0, le=360)
    product_name: str | None = Field(default=None, max_length=180)


class SafeInstallmentSuggestion(BaseModel):
    installments: int
    installment_value: Decimal
    monthly_impact_percentage: Decimal
    risk_level: RiskLevel


class CanIBuyResponse(BaseModel):
    approved: bool
    risk_level: RiskLevel
    monthly_impact_percentage: Decimal
    new_installment_value: Decimal
    remaining_balance_after_purchase: Decimal
    salary: Decimal
    recurring_expenses: Decimal
    current_installment_commitment: Decimal
    total_committed_after_purchase: Decimal
    financial_health_score: int
    recommendation: str
    warnings: list[str]
    safe_installment_suggestions: list[SafeInstallmentSuggestion]

File: frontend/src/types/financial.ts
export type RiskLevel = "low" | "medium" | "high";

export interface MonthSummary {
  year: number;
  month: number;
  salary: string;
  total_expenses: string;
  monthly_installment_commitment: string;
  total_committed: string;
  remaining_balance: string;
  committed_percentage: string;
  active_installments: number;
}

export interface FutureMonthBalance {
  months_ahead: number;
  year: number;
  month: number;
  projected_installment_commitment: string;
  projected_balance: string;
  active_installments: number;
}

export interface FutureBalanceResponse {
  salary: string;
  months: FutureMonthBalance[];
}

export interface CanIBuyRequest {
  product_price: number;
  installments: number;
  product_name?: string;
}

export interface SafeInstallmentSuggestion {
  installments: number;
  installment_value: string;
  monthly_impact_percentage: string;
  risk_level: RiskLevel;
}

export interface CanIBuyResponse {
  approved: boolean;
  risk_level: RiskLevel;
  monthly_impact_percentage: string;
  new_installment_value: string;
  remaining_balance_after_purchase: string;
  salary: string;
  recurring_expenses: string;
  current_installment_commitment: string;
  total_committed_after_purchase: string;
  financial_health_score: number;
  recommendation: string;
  warnings: string[];
  safe_installment_suggestions: SafeInstallmentSuggestion[];
}
